### ISSUE TITLE
`prefer_constructors_over_static_methods` false positives

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#2330]
See: https://github.com/dart-lang/linter/pull/2328#discussion_r517104666

**Comment by pq** [comment]
Looking at this a little closer, I don't think the lint ever worked as intended.  Notably, given a class `A` with a static method `ok()` we are not checking that `ok()` returns `A` or a subtype. 

The fact that we report here seems incorrect:

```dart
class A { }

class C {
  static A ok() => A(); // false positive
}


```

**Comment by pq** [comment]
Moreover, as @bwilkerson points out, in the presence of nullability:

> The lint needs to not flag static methods from which null can be returned if null safety is enabled. Under null safety it is no longer possible to return null from a factory constructor.

**Comment by pq** [comment]
Assignable types addressed w/ #2333.  NNBD case is tracked in #2280.
